src/main/scala/mesosphere/marathon/core/storage/repository/impl/DeploymentRepositoryImpl.scala

```diff
@@ -97,4 +120,169 @@ class DeploymentRepositoryImpl[K, C, S](
         None
     }
   }
+
+  private[impl] def lazyAll(): Source[StoredPlan, NotUsed] =
+    repo.ids().mapAsync(Int.MaxValue)(repo.get).collect { case Some(g) => g }
+}
+
+/**
+  * Actor  that when a plan is deleted will:
+  *
+  * - start up a deletion, future deletes are coalesced into the next delete "round"
+  * - Look at all the current deployment plans + all of their roots and the current root.
+  * - Look at all of the root versions, if there are more than max versions, delete
+  *   the oldest root that is not referred to by any other plan.
+  * - If there are roots to delete, delete them, then go through all of the apps and app versions
+  *   from _all of the roots currently referenced_ and find appVersions that are either
+  *   (a) able to be _completely removed_
+  *   (b) have too many versions
+  * - Apps that are referenced by _no roots_ are deleted.
+  * - Apps that have too many versions will be pruned by deleting the oldest versions
+  *   that are not referenced by any roots.
+  *
+  *
+  * @todo I believe we technically need transactions or we need to some how block
+  *       apps/roots from being stored _while_ GC is taking place.
+  */
+private class GcActor[K, C, S](
+  deploymentRepository: DeploymentRepositoryImpl[K, C, S],
+  groupRepository: StoredGroupRepositoryImpl[K, C, S],
+  appRepository: AppRepositoryImpl[K, C, S],
+  maxVersions: Int)(implicit mat: Materializer, ctx: ExecutionContext)
+    extends Actor with StrictLogging {
+  override val receive: Receive = idle
+
+  def idle: Receive = {
+    case DeletedPlan(plan) =>
+      delete()
+      context.become(deleting(Nil))
+  }
+
+  def deleting(pending: List[DeploymentPlan]): Receive = {
+    case DeletedPlan(plan) =>
+      context.become(deleting(plan :: pending))
+    case DeleteDone =>
+      if (pending.nonEmpty) {
+        delete()
+        context.become(deleting(Nil))
```

Comment:
You are not looking into the pending List. If delete() will review all nodes, why do we have a list here?